packages/backend/src/modules/tvl/services/AggLayerService.ts

```diff
@@ -0,0 +1,210 @@
+import { Logger } from '@l2beat/backend-tools'
+import { AmountRecord } from '@l2beat/database'
+import {
+  AggLayerL2Token,
+  AggLayerNativeEtherPreminted,
+  AggLayerNativeEtherWrapped,
+  Bytes,
+  EthereumAddress,
+  UnixTime,
+} from '@l2beat/shared-pure'
+import { BigNumber, utils } from 'ethers'
+import { MulticallClient } from '../../../peripherals/multicall/MulticallClient'
+import { MulticallRequest } from '../../../peripherals/multicall/types'
+import { RpcClient } from '../../../peripherals/rpcclient/RpcClient'
+
+export const erc20Interface = new utils.Interface([
+  'function balanceOf(address account) view returns (uint256)',
+  'function totalSupply() view returns (uint256)',
+])
+
+export const bridgeInterface = new utils.Interface([
+  'function getTokenWrappedAddress(uint32 originNetwork, address originTokenAddress) view returns (address)',
+])
+
+export const BRIDGE_ADDRESS = EthereumAddress(
+  '0x2a3DD3EB832aF982ec71669E178424b10Dca2EDe',
+)
+
+type AggLayerAmountConfig =
+  | AggLayerL2Token
+  | AggLayerNativeEtherPreminted
+  | AggLayerNativeEtherWrapped
+
+export type Config = AggLayerAmountConfig & { id: string }
+
+export interface AggLayerServiceDependencies {
+  readonly rpcClient: RpcClient
+  readonly multicallClient: MulticallClient
+  logger: Logger
+}
+
+export class AggLayerService {
+  constructor(private readonly $: AggLayerServiceDependencies) {
+    this.$.logger = $.logger.for(this)
+  }
+
+  async fetchAmounts(
+    tokens: (Config & { type: 'agglayerL2Token' })[],
```

Comment:
I do not get those unions "Config & {type...}", wouldnt it be simpler to have just the type & {configId} ?

Reply:
to reduce boilerplate even more you can create generic type for it